File: frontend/flutter/lib/utils/schedule_engine.dart
import '../models/timetable_entry.dart';
import '../models/app_settings.dart';

class ScheduleEngine {
  final AppSettings config;

  ScheduleEngine(this.config);

  int timeToMinutes(String t) {
    final parts = t.split(':');
    return int.parse(parts[0]) * 60 + int.parse(parts[1]);
  }

  String minutesToTime(int m) {
    final h = m ~/ 60;
    final mn = m % 60;
    return '${h.toString().padLeft(2, '0')}:${mn.toString().padLeft(2, '0')}';
  }

  List<Map<String, dynamic>> generateDailySlots() {
    final slots = <Map<String, dynamic>>[];
    int current = timeToMinutes(config.schoolStart);
    final endDay = timeToMinutes('17:00');
    int periodNum = 1;

    // Build break intervals
    final breakIntervals = config.breakSlots.map((b) => {
      'name': b.name,
      'start': timeToMinutes(b.start),
      'end': timeToMinutes(b.start) + b.duration,
    }).toList();

    while (current < endDay && periodNum <= config.numPeriods) {
      // Check if current time falls within a break
      final inBreak = breakIntervals.cast<Map<String, dynamic>>().firstWhere(
        (b) => current >= (b['start'] as int) && current < (b['end'] as int),
        orElse: () => <String, dynamic>{},
      );

      if (inBreak.isNotEmpty) {
        slots.add({
          'type': 'break',
          'name': inBreak['name'],
          'startTime': minutesToTime(inBreak['start'] as int),
          'endTime': minutesToTime(inBreak['end'] as int),
          'periodNumber': null,
        });
        current = inBreak['end'] as int;
        continue;
      }

      // Check if slot would overlap with break
      final nextBreak = breakIntervals.cast<Map<String, dynamic>>().firstWhere(
        (b) => current < (b['end'] as int) && current + config.periodDuration > (b['start'] as int),
        orElse: () => <String, dynamic>{},
      );

      int slotEnd;
      if (nextBreak.isNotEmpty) {
        slotEnd = nextBreak['start'] as int;
      } else {
        slotEnd = (current + config.periodDuration) > endDay ? endDay : current + config.periodDuration;
      }

      final duration = slotEnd - current;
      if (duration > 0) {
        slots.add({
          'type': 'teaching',
          'periodNumber': periodNum,
          'startTime': minutesToTime(current),
          'endTime': minutesToTime(slotEnd),
          'duration': duration,
        });
        periodNum++;
      }
      current = slotEnd;
    }
    return slots;
  }

  TimetableEntry? getCurrentPeriod(List<TimetableEntry> assignments, DateTime now) {
    const days = ['SUN', 'M', 'T', 'W', 'Th', 'F', 'S'];
    final dayKey = days[now.weekday % 7];
    if (dayKey == 'SUN' || dayKey == 'S') return null;

    final nowMinutes = now.hour * 60 + now.minute;
    final todayAssignments = assignments.where((a) => a.day == dayKey).toList();

    for (final a in todayAssignments) {
      if (a.startTime.isNotEmpty) {
        final start = timeToMinutes(a.startTime);
        final end = start + (a.duration);
        if (nowMinutes >= start && nowMinutes < end) return a;
      }
    }
    return null;
  }

  TimetableEntry? getNextPeriod(List<TimetableEntry> assignments, DateTime now) {
    const days = ['SUN', 'M', 'T', 'W', 'Th', 'F', 'S'];
    final dayKey = days[now.weekday % 7];
    if (dayKey == 'SUN' || dayKey == 'S') return null;

    final nowMinutes = now.hour * 60 + now.minute;
    final todayAssignments = assignments
        .where((a) => a.day == dayKey)
        .toList()
      ..sort((a, b) => timeToMinutes(a.startTime).compareTo(timeToMinutes(b.startTime)));

    for (final a in todayAssignments) {
      if (a.startTime.isNotEmpty) {
        final start = timeToMinutes(a.startTime);
        if (nowMinutes < start) return a;
      }
    }
    return null;
  }

  List<Map<String, dynamic>> detectConflicts(List<TimetableEntry> assignments) {
    final conflicts = <Map<String, dynamic>>[];
    final daySlots = <String, List<TimetableEntry>>{};

    for (final a in assignments) {
      daySlots.putIfAbsent(a.day, () => []).add(a);
    }

    for (final entry in daySlots.entries) {
      final slots = entry.value;
      for (int i = 0; i < slots.length; i++) {
        for (int j = i + 1; j < slots.length; j++) {
          final a = slots[i];
          final b = slots[j];
          final aStart = timeToMinutes(a.startTime);
          final aEnd = aStart + a.duration;
          final bStart = timeToMinutes(b.startTime);
          final bEnd = bStart + b.duration;
          if (aStart < bEnd && bStart < aEnd) {
            conflicts.add({
              'day': entry.key,
              'subject1': a.subject,
              'subject2': b.subject,
              'time1': a.startTime,
              'time2': b.startTime,
              'message': 'Overlap on ${entry.key}: "${a.subject}" (${a.startTime}) and "${b.subject}" (${b.startTime})',
            });
          }
        }
      }
    }
    return conflicts;
  }

  Map<String, double> computeTeachingHours(List<TimetableEntry> assignments) {
    final hours = <String, double>{};
    for (final a in assignments) {
      if (a.subject.isNotEmpty) {
        hours[a.subject] = (hours[a.subject] ?? 0) + (a.duration / 60.0);
      }
    }
    return hours;
  }
}

File: backend/utils/scheduleEngine.js
class ScheduleEngine {
  constructor(config = {}) {
    this.schoolStart = config.schoolStart || '08:00';
    this.periodDuration = config.periodDuration || 40;
    this.breakSlots = config.breakSlots || [
      { name: 'Break', start: '10:00', duration: 20 },
      { name: 'Lunch', start: '11:40', duration: 90 },
      { name: 'Afternoon Break', start: '15:10', duration: 20 },
    ];
    this.numPeriods = config.numPeriods || 12;
  }

  timeToMinutes(t) {
    const [h, m] = t.split(':').map(Number);
    return h * 60 + m;
  }

  minutesToTime(m) {
    const h = Math.floor(m / 60);
    const mn = m % 60;
    return `${String(h).padStart(2, '0')}:${String(mn).padStart(2, '0')}`;
  }

  generateDailySlots() {
    const slots = [];
    let current = this.timeToMinutes(this.schoolStart);
    const endDay = this.timeToMinutes('17:00');
    let periodNum = 1;

    // Build break intervals in minutes
    const breakIntervals = this.breakSlots.map(b => ({
      name: b.name,
      start: this.timeToMinutes(b.start),
      end: this.timeToMinutes(b.start) + b.duration,
    }));

    while (current < endDay && periodNum <= this.numPeriods) {
      // Check if current time falls within a break
      const inBreak = breakIntervals.find(b => current >= b.start && current < b.end);
      if (inBreak) {
        slots.push({
          type: 'break',
          name: inBreak.name,
          startTime: this.minutesToTime(inBreak.start),
          endTime: this.minutesToTime(inBreak.end),
          duration: inBreak.end - inBreak.start,
          periodNumber: null,
        });
        current = inBreak.end;
        continue;
      }

      // Check if slot would overlap with break
      const nextBreak = breakIntervals.find(b => current < b.end && current + this.periodDuration > b.start);
      let slotEnd;
      if (nextBreak) {
        slotEnd = nextBreak.start;
      } else {
        slotEnd = Math.min(current + this.periodDuration, endDay);
      }

      const duration = slotEnd - current;
      if (duration > 0) {
        slots.push({
          type: 'teaching',
          periodNumber: periodNum,
          startTime: this.minutesToTime(current),
          endTime: this.minutesToTime(slotEnd),
          duration: duration,
        });
        periodNum++;
      }
      current = slotEnd;
    }
    return slots;
  }

  detectConflicts(assignments) {
    const conflicts = [];
    const daySlots = { M: [], T: [], W: [], Th: [], F: [] };

    for (const a of assignments) {
      daySlots[a.day] = daySlots[a.day] || [];
      daySlots[a.day].push(a);
    }

    for (const [day, slots] of Object.entries(daySlots)) {
      for (let i = 0; i < slots.length; i++) {
        for (let j = i + 1; j < slots.length; j++) {
          const a = slots[i];
          const b = slots[j];
          if (a.startTime && a.endTime && b.startTime && b.endTime) {
            const aStart = this.timeToMinutes(a.startTime);
            const aEnd = aStart + (a.duration || 40);
            const bStart = this.timeToMinutes(b.startTime);
            const bEnd = bStart + (b.duration || 40);
            if (aStart < bEnd && bStart < aEnd) {
              conflicts.push({
                day,
                subject1: a.subject,
                subject2: b.subject,
                time1: a.startTime,
                time2: b.startTime,
                message: `Overlap on ${day}: "${a.subject}" (${a.startTime}) and "${b.subject}" (${b.startTime})`,
              });
            }
          }
        }
      }
    }
    return conflicts;
  }

  getCurrentPeriod(assignments, now = new Date()) {
    const days = ['SUN', 'M', 'T', 'W', 'Th', 'F', 'S'];
    const dayKey = days[now.getDay()];
    if (dayKey === 'SUN' || dayKey === 'S') return null;

    const nowMinutes = now.getHours() * 60 + now.getMinutes();
    const todayAssignments = assignments.filter(a => a.day === dayKey);

    for (const a of todayAssignments) {
      if (a.startTime && a.endTime) {
        const start = this.timeToMinutes(a.startTime);
        const end = start + (a.duration || 40);
        if (nowMinutes >= start && nowMinutes < end) {
          return { ...a, remaining: end - nowMinutes };
        }
      }
    }
    return null;
  }

  getNextPeriod(assignments, now = new Date()) {
    const days = ['SUN', 'M', 'T', 'W', 'Th', 'F', 'S'];
    const dayKey = days[now.getDay()];
    if (dayKey === 'SUN' || dayKey === 'S') return null;

    const nowMinutes = now.getHours() * 60 + now.getMinutes();
    const todayAssignments = assignments
      .filter(a => a.day === dayKey)
      .sort((a, b) => this.timeToMinutes(a.startTime) - this.timeToMinutes(b.startTime));

    for (const a of todayAssignments) {
      if (a.startTime) {
        const start = this.timeToMinutes(a.startTime);
        if (nowMinutes < start) {
          return { ...a, minutesUntil: start - nowMinutes };
        }
      }
    }
    return null;
  }

  computeTeachingHours(assignments) {
    const hours = {};
    for (const a of assignments) {
      if (a.subject) {
        hours[a.subject] = (hours[a.subject] || 0) + (a.duration || 40);
      }
    }
    // Convert minutes to hours
    for (const k of Object.keys(hours)) {
      hours[k] = Math.round((hours[k] / 60) * 10) / 10;
    }
    return hours;
  }
}

module.exports = ScheduleEngine;
